packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,308 @@
+        );
+        assertValidFill(
+            rightOrder,
+            rightOrderInfo.orderStatus,
+            rightOrderInfo.orderHash,
+            takerAddress,
+            rightOrderInfo.orderFilledAmount,
+            matchedFillResults.right.takerAssetFilledAmount,
+            rightSignature
+        );
+
+        // Settle matched orders. Succeeds or throws.
+        settleMatchedOrders(
+            leftOrder,
+            rightOrder,
+            takerAddress,
+            matchedFillResults
+        );
+
+        // Update exchange state
+        updateFilledState(
+            leftOrder,
+            takerAddress,
+            leftOrderInfo.orderHash,
+            leftOrderInfo.orderFilledAmount,
+            matchedFillResults.left
+        );
+        updateFilledState(
+            rightOrder,
+            takerAddress,
+            rightOrderInfo.orderHash,
+            rightOrderInfo.orderFilledAmount,
+            matchedFillResults.right
+        );
+
+        return matchedFillResults;
+    }
+
+    /// @dev Validates context for matchOrders. Succeeds or throws.
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    function assertValidMatch(
+        Order memory leftOrder,
+        Order memory rightOrder
+    )
+        internal
+    {
+        // The leftOrder maker asset must be the same as the rightOrder taker asset.
+        require(
+            areBytesEqual(leftOrder.makerAssetData, rightOrder.takerAssetData),
+            ASSET_MISMATCH_MAKER_TAKER
+        );
+
+        // The leftOrder taker asset must be the same as the rightOrder maker asset.
+        require(
+            areBytesEqual(leftOrder.takerAssetData, rightOrder.makerAssetData),
+            ASSET_MISMATCH_TAKER_MAKER
+        );
+
+        // Make sure there is a positive spread.
+        // There is a positive spread iff the cost per unit bought (OrderA.MakerAmount/OrderA.TakerAmount) for each order is greater
+        // than the profit per unit sold of the matched order (OrderB.TakerAmount/OrderB.MakerAmount).
+        // This is satisfied by the equations below:
+        // <leftOrder.makerAssetAmount> / <leftOrder.takerAssetAmount> >= <rightOrder.takerAssetAmount> / <rightOrder.makerAssetAmount>
+        // AND
+        // <rightOrder.makerAssetAmount> / <rightOrder.takerAssetAmount> >= <leftOrder.takerAssetAmount> / <leftOrder.makerAssetAmount>
+        // These equations can be combined to get the following:
+        require(
+            safeMul(leftOrder.makerAssetAmount, rightOrder.makerAssetAmount) >=
+            safeMul(leftOrder.takerAssetAmount, rightOrder.takerAssetAmount),
+            NEGATIVE_SPREAD
+        );
+    }
+
+    /// @dev Validates matched fill results. Succeeds or throws.
+    /// @param matchedFillResults Amounts to fill and fees to pay by maker and taker of matched orders.
+    function assertValidMatch(MatchedFillResults memory matchedFillResults)
```

Comment:
I'd prefer not to overload this function. How about `assertValidMatchOrders` and `assertValidMatchResults`?